src/intrusive_double_linked_list.rs

```diff
@@ -48,278 +50,335 @@ impl<T> DerefMut for ListNode<T> {
+        node.next = None;
+        self.tail.as_mut().unwrap().as_mut().next = Some(node.into());
+        self.tail = Some(node.into());
     }
 
-    /// Returns the first item in the linked list without removing it from the list
+    /// Returns the first node in the linked list without removing it from the list
     /// The function is only safe as long as valid pointers are stored inside
     /// the linked list.
     /// The returned pointer is only guaranteed to be valid as long as the list
     /// is not mutated
-    pub fn peek_first(&self) -> *mut ListNode<T> {
-        self.head
+    pub fn peek_first(&self) -> Option<&mut ListNode<T>> {
+        // Safety: When the node was inserted it was promised that it is alive
+        // until it gets removed from the list.
+        // The returned node has a pointer which constrains it to the lifetime
+        // of the list. This is ok, since the Node is supposed to outlive
+        // its insertion in the list.
+        unsafe {
+            self.head
+                .map(|mut node| &mut *(node.as_mut() as *mut ListNode<T>))
+        }
     }
 
-    /// Returns the last item in the linked list without removing it from the list
+    /// Returns the last node in the linked list without removing it from the list
     /// The function is only safe as long as valid pointers are stored inside
     /// the linked list.
     /// The returned pointer is only guaranteed to be valid as long as the list
     /// is not mutated
-    pub fn peek_last(&self) -> *mut ListNode<T> {
-        self.tail
-    }
-
-    /// Removes the last item from the linked list and returns it
-    /// Safety: This function is only safe as long as all pointers which are
-    /// stored inside the linked list are valid.
-    pub unsafe fn remove_last(&mut self) -> *mut ListNode<T> {
-        if self.tail.is_null() {
-            return null_mut();
+    pub fn peek_last(&self) -> Option<&mut ListNode<T>> {
+        // Safety: When the node was inserted it was promised that it is alive
+        // until it gets removed from the list.
+        // The returned node has a pointer which constrains it to the lifetime
+        // of the list. This is ok, since the Node is supposed to outlive
+        // its insertion in the list.
+        unsafe {
+            self.tail
+                .map(|mut node| &mut *(node.as_mut() as *mut ListNode<T>))
         }
+    }
 
-        let last = self.tail;
-        self.tail = (*last).prev;
-        if !(*last).prev.is_null() {
-            (*(*last).prev).next = null_mut();
-        } else {
-            // This was the last item in the list
-            self.head = null_mut();
+    /// Removes the first node from the linked list
+    pub fn remove_first(&mut self) -> Option<&mut ListNode<T>> {
+        // Safety: When the node was inserted it was promised that it is alive
+        // until it gets removed from the list
+        unsafe {
+            match self.head {
```

Comment:
nit/TIOLI: instead of matching, you could just say
```rust
let mut head = self.head?;

self.head = head.as_mut().next;

let first_ref = head.as_mut();

// ... rest of method
```

Reply:
That's indeed better! Forgot that `?` actually also works for Options.